src/metrics/rules/active_ruleset.go

```diff
@@ -310,28 +332,31 @@ func (as *activeRuleSet) rollupResultsFor(id []byte, timeNanos int64, matchOpts
 // nolint: unparam
 func (as *activeRuleSet) toRollupResults(
 	id []byte,
+	timeNanos int64,
 	cutoverNanos int64,
 	targets []rollupTarget,
 	keepOriginal bool,
 	tags [][]models.Tag,
-	matchOpts MatchOptions,
-) (rollupResults, error) {
+	matchOpts *MatchOptions,
+) error {
 	if len(targets) == 0 {
-		return rollupResults{}, nil
+		return nil
 	}
 
 	// If we cannot extract tags from the id, this is likely an invalid
 	// metric and we bail early.
 	_, sortedTagPairBytes, err := matchOpts.NameAndTagsFn(id)
 	if err != nil {
-		return rollupResults{}, err
+		return err
 	}
 
 	var (
-		multiErr           = xerrors.NewMultiError()
-		pipelines          = make([]metadata.PipelineMetadata, 0, len(targets))
-		newRollupIDResults = make([]idWithMatchResults, 0, len(targets))
-		tagPairs           []metricid.TagPair
+		multiErr    = xerrors.NewMultiError()
+		tagPairs    []metricid.TagPair
```

Comment:
i wonder how bad `tagPairs` is. might be worth a follow up

Reply:
I remember seeing it on the profile, so it's not great. We definitely should do this one.

Reply:
yea probably a good follow up (or even in this PR) is to remove all heap allocations from ForwardMatch, since it's in the hot path O(metrics * rules)